Fix bug in a unit test

jdavcs opened the pull request:
There was a bug in a context manager inside a unit test. This caused
nondeterministic behavior of some unit tests after a refactoring.

Bug fixed; added 3 unit tests to test the context manager.

Ping @jmchilton @natefoo

jdavcs committed:
Fix bug in a unit test

There was a bug in a context manager inside a unit test. This caused
nondeterministic behavior of some unit tests after a refactoring.

Bug fixed; added 3 unit tests to test the context manager.

jdavcs committed:
Fix lint errors

natefoo reviewed (approved):
Good catch!

mvdbeek commented on test/unit/tools/test_tool_deps.py:
```diff
@@ -719,25 +719,69 @@ def _first_conda_resolver_options(dm):
 
 @contextmanager
-def __environ(values, remove=[]):
+def __environ(values, keys_to_remove=[]):
```
I distribute bonus points for eliminating mutable default args, see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments. The way you use it here it's fine,
but better to just say `keys_to_remove=None` and then 
```
if keys_to_remove is None:
    keys_to_remove = []
```
in the function body.

jdavcs replied:
@mvdbeek Yes! (and there's no need to create [] in the body: if it's none, we never use it - so a simple check will be sufficient). Thx!

jdavcs replied:
Done.

nsoranzo commented on test/unit/tools/test_tool_deps.py:
```diff
@@ -719,25 +719,69 @@ def _first_conda_resolver_options(dm):
+def __environ(values, keys_to_remove=[]):
     """
     Modify the environment for a test, adding/updating values in dict `values` and
     removing any environment variables mentioned in list `remove`.
```
```suggestion
    removing any environment variables mentioned in list `keys_to_remove`.
```

jdavcs replied:
@nsoranzo Of course!

jdavcs replied:
Done.

nsoranzo commented on test/unit/tools/test_tool_deps.py:
```diff
@@ -719,25 +719,69 @@ def _first_conda_resolver_options(dm):
+    new_val2 = '43'  # update old key's value
+    to_update = {new_key: new_val1, key: new_val2}
+    with __environ(values=to_update):
+        pass
```
What about testing that `os.environ` has been updated here?

jdavcs replied:
Done!

nsoranzo commented on test/unit/tools/test_tool_deps.py:
```diff
@@ -719,25 +719,69 @@ def _first_conda_resolver_options(dm):
+    new_key, new_val = 'foo_test_new', '53'  # add new key + value
+    to_update = {new_key: new_val}
+    with __environ(values=to_update, keys_to_remove=[key, new_key]):
+        pass
```
Same here.

jdavcs replied:
Done!

martenson commented:
switching to wip until the comments are addressed

jdavcs committed:
Merge remote-tracking branch 'upstream/dev' into dev_unittest_bugfix

jdavcs committed:
Restructure/modify environ context manager + tests

Factor out __environ context manager into helper class (and rename).

Move context manager tests into separate test module. Improve tests.

Refactor context manager:
- Change default arg to None
- Simplify: I don't think we need to keep track of new keys;
  environ.clear() + environ.update(old_environ) should be sufficient.
  We significantly reduce complexity at the minor cost of a little more
  space.
- Instead of wrapping in a try/except block, just check if key exists
  before deleting.

Modify test_tool_deps to use the new context manager.

jdavcs commented:
I have made several modifications to the context manager + tests (also, see notes in latest commit). As far as I can tell, the main change to the context manager (no more new_keys) shouldn't cause any issues. Please take a look. @nsoranzo @mvdbeek

jdavcs committed:
Fix linting errors (import stms)

jdavcs committed:
Fix package test

The cause of this is related to pulling tool_util out of tools:
https://github.com/galaxyproject/galaxy/pull/8279

The solution here is a temporary hack (add helper module, util.py, to
tests/, then make tests a package).

A better solution will be to separate test/unit/tools into tools/ and
tool_util/ - I intend to take care of it soon.

jdavcs commented:
E: Unable to locate package libpython3.5-dev??
Can someone rerun the tests, please?

jmchilton commented:
Looks good to me, that bug with Python 3.5 is fixed upstream so I wouldn't worry about it. Take out of WIP?

jdavcs commented:
@jmchilton thanks! Yes, all good to go then.

jdavcs committed:
Merge remote-tracking branch 'upstream/dev' into dev_unittest_bugfix

jdavcs commented:
This one is new: OperationalError: (psycopg2.OperationalError) FATAL:  remaining connection slots are reserved for non-replication superuser connections. It's unrelated to this PR (as far as I can judge). Will rerun tests.

jdavcs committed:
Merge remote-tracking branch 'upstream/dev' into dev_unittest_bugfix

jdavcs commented:
Ready for merging!
